File: python/剑指offer/面试题59 - I. 滑动窗口的最大值.py
from typing import List

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        if nums == [] or k == 0:
            return []
        result = []
        i = 0
        while i < len(nums)-k+1:
            # print(nums[i:i+k])
            result.append(max(nums[i:i+k]))
            i+=1
        return result


# 双向队列
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        if not nums:
            return []
        from collections import deque
        queue = deque()
        i = 0
        while i < k:
            queue.append(nums[i])
            i+=1
        res = []
        res.append(max(queue))
        while i < len(nums):
            queue.popleft()
            queue.append(nums[i])
            res.append(max(queue))
            i+=1
        return res


# 单调队列
class Solution3:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        import collections
        if not nums or k == 0: return []
        deque = collections.deque()
        for i in range(k): # 未形成窗口
            while deque and deque[-1] < nums[i]: deque.pop()
            deque.append(nums[i])
        res = [deque[0]]
        for i in range(k, len(nums)): # 形成窗口后
            if deque[0] == nums[i - k]: deque.popleft()
            while deque and deque[-1] < nums[i]: deque.pop()
            deque.append(nums[i])
            res.append(deque[0])
        return res

s = Solution3()
nums = [1,3,-1,-3,5,3,6,7]
k = 3
print(s.maxSlidingWindow(nums,k))

File: go/剑指offer/面试题59 - I. 滑动窗口的最大值.go
package main

func maxSlidingWindow(nums []int, k int) []int {
	if len(nums) == 0 {
		return []int{}
	}

	deque := make([]int, 0, k)
	for i := 0; i < k; i++ {
		for len(deque) > 0 && deque[len(deque)-1] < nums[i] {
			deque = deque[:len(deque)-1]
		}
		deque = append(deque, nums[i])
	}

	maxs := make([]int, 0, len(nums)-k+1)
	maxs = append(maxs, deque[0])
	for i := k; i < len(nums); i++ {
		if nums[i-k] == deque[0] {
			deque = deque[1:]
		}
		for len(deque) > 0 && deque[len(deque)-1] < nums[i] {
			deque = deque[:len(deque)-1]
		}
		deque = append(deque, nums[i])
		maxs = append(maxs, deque[0])
	}
	return maxs
}
